server/src/main/java/org/opensearch/tasks/CancellableTask.java

```diff
@@ -47,17 +48,17 @@
  */
 public abstract class CancellableTask extends Task {
 
-    private volatile String reason;
+    private final SetOnce<String> reason = new SetOnce<>();
     private final AtomicBoolean cancelled = new AtomicBoolean(false);
     private final TimeValue cancelAfterTimeInterval;
     /**
      * The time this task was cancelled as a wall clock time since epoch ({@link System#currentTimeMillis()} style).
      */
-    private Long cancellationStartTime = null;
+    private final SetOnce<Long> cancellationStartTime = new SetOnce<>();
     /**
      * The time this task was cancelled as a relative time ({@link System#nanoTime()} style).
      */
-    private Long cancellationStartTimeNanos = null;
+    private final SetOnce<Long> cancellationStartTimeNanos = new SetOnce<>();
```

Comment:
To clarify, I didn't suggest using SetOnce for all these properties.

Our goal is to make updates to these properties all at once (i.e. atomic). To achieve this, we need to wrap all these properties in a single object, and SetOnce that object atomically. Here's what it could look like:

```java
public abstract class CancellableTask extends Task {

    // ... truncated

    static class CancelledInfo {
        String reason;
        Long cancellationStartTime;
        Long cancellationStartTimeNanos;

        public CancelledInfo(String reason) {
            this.reason = reason;
            this.cancellationStartTime = System.currentTimeMillis();
            this.cancellationStartTimeNanos = System.nanoTime();
        }
    }

    private final SetOnce<CancelledInfo> cancelledInfo = new SetOnce<>();

    public void cancel(String reason) {
        assert reason != null;
        if (cancelledInfo.trySet(new CancelledInfo(reason))) {
            onCancelled();
        }
    }

    public boolean isCancelled() {
        return cancelledInfo.get() != null;
    }

    @Nullable
    public final String getReasonCancelled() {
        CancelledInfo info = cancelledInfo.get();
        return (info != null) ? info.reason : null;
    }
}
```

Reply:
My apologies. 

I see I misunderstood your previous suggestion. Since I could not replicate the issue in a reasonable number of test trials I figured the issue had been resolved.

Reply:
Please check my comment above to reliably reproduce this issue.